contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol

```diff
@@ -40,18 +38,26 @@ library ChainSpecificUtil {
 
   // ------------ End Optimism Constants ------------
 
-  /// @notice Returns the L1 fees in wei that will be paid for the current transaction, given any calldata
-  /// @notice for the current transaction.
-  /// @notice When on a known Arbitrum chain, it uses ArbGas.getCurrentTxL1GasFees to get the fees.
-  /// @notice On Arbitrum, the provided calldata is not used to calculate the fees.
-  /// @notice On Optimism, the provided calldata is passed to the GasPriceOracle predeploy
-  /// @notice and getL1Fee is called to get the fees.
-  function _getCurrentTxL1GasFees(bytes memory txCallData) internal view returns (uint256 l1FeeWei) {
+  /// @notice Returns the upper limit estimate of the L1 fees in wei that will be paid for L2 chains
+  /// @notice based on the size of the transaction data and the current gas conditions.
+  /// @notice This is an "upper limit" as it assumes the transaction data is uncompressed when posted on L1.
+  function _getL1FeeUpperLimit(uint256 calldataSizeBytes) internal view returns (uint256 l1FeeWei) {
     uint256 chainid = block.chainid;
     if (_isArbitrumChainId(chainid)) {
-      return ARBGAS.getCurrentTxL1GasFees();
+      // https://docs.arbitrum.io/build-decentralized-apps/how-to-estimate-gas#where-do-we-get-all-this-information-from
+      (, uint256 l1PricePerByte, , , , ) = ARBGAS.getPricesInWei();
+      return l1PricePerByte * (calldataSizeBytes + DATA_PADDING_SIZE);
     } else if (_isOptimismChainId(chainid)) {
-      return OVM_GASPRICEORACLE.getL1Fee(bytes.concat(txCallData, L1_FEE_DATA_PADDING));
+      // https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#ecotone-l1-cost-fee-changes-eip-4844-da
+      // note we conservatively assume all non-zero bytes: tx_compressed_size = tx_data_size_bytes
+      uint256 l1BaseFeeWei = L1BLOCK.basefee();
+      uint256 l1BaseFeeScalar = L1BLOCK.baseFeeScalar();
+      uint256 l1BlobBaseFeeWei = L1BLOCK.blobBaseFee();
+      uint256 l1BlobBaseFeeScalar = L1BLOCK.blobBaseFeeScalar();
+      return
+        ((calldataSizeBytes + DATA_PADDING_SIZE) * // full tx data size in bytes. Includes additional 140 bytes of "static data" (signature, etc.)
+          (16 * l1BaseFeeScalar * l1BaseFeeWei + l1BlobBaseFeeScalar * l1BlobBaseFeeWei)) / (10 ** 6);
```

Comment:
I think you are missing division by 16? This is stated in the [GitHub link](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#ecotone-l1-cost-fee-changes-eip-4844-da) above and also [here](https://docs.optimism.io/stack/transactions/fees#ecotone)

Also, can you explain why the data padding size is 140 bytes? I'm looking at the [OracleGasPrice implementation details](https://optimistic.etherscan.io/address/0xb528d11cc114e026f138fe568744c6d45ce6da7a#code) and it looks like they are taking only [68 bytes padding](https://github.com/ethereum-optimism/optimism/blob/233ede59d16cb01bdd8e7ff662a153a4c3178bdd/packages/contracts-bedrock/contracts/L2/GasPriceOracle.sol#L110) for the transaction signature into account.

Reply:
For example, this is my code based on the official OP GasPriceOracle implementation:
```
  function _calculateOptimismL1DataFee(uint256 calldataSizeBytes) internal view returns (uint256) {
    // 68 bytes of padding for transaction signature
    uint256 l1GasUsed = calldataSizeBytes * 16 + (68 * 16);
    uint256 scaledBaseFee = L1BLOCK.baseFeeScalar() * 16 * OVM_GASPRICEORACLE.l1BaseFee();
    uint256 scaledBlobBaseFee = L1BLOCK.blobBaseFeeScalar() * L1BLOCK.blobBaseFee();
    uint256 fee = l1GasUsed * (scaledBaseFee + scaledBlobBaseFee);
    return fee / (16 * 10 ** OVM_GASPRICEORACLE.decimals());
  }
```

Reply:
> Does this make sense? Please double-check carefully as I may be incorrect.

@KuphJr I checked and it makes sense 👍 @matYang explained it very well in the comment above
I didn't get to the point of calculating the correct padding size yet for VRF fulfillment, but I do want to run some tests and make comparisons. That's why I was interested in understanding how 140 bytes was selected as the padding size.
If the OP team gives you any further clarification on this topic, do let me know!

Reply:
if calldataSizeBytes is the unsigned transaction length then you'll want to add 68 bytes for the signature, not 140.

Reply:
Just updated the PR!
I think that should be the final commit needed.